update: improve Classes page (#4786)
JetBrains/kotlin-web-site: docs/topics/classes.md

To declare a primary constructor, use the class header after the class name:

```kotlin
class Person constructor(name: String) { /*...*/ }
```

In this example, the primary constructor initializes the class `Person` by defining a parameter `name` of type `String`.

If the primary constructor does not have any annotations or [visibility modifiers](visibility-modifiers.md#constructors), 
you can omit the `constructor` keyword:

```kotlin
class Person(name: String) { /*...*/ }
```

You can define simple properties directly in the primary constructor. To declare a read-only property, 
use the `val` keyword in the parentheses before the argument name. 
For a mutable property, use the `var` keyword:

```kotlin
class Person(val name: String, var age: Int) { /*...*/ }
```

In this example, the `Person` class has a primary constructor with two properties: `name` and `age`. 
The `name` property is declared as read-only using `val`, while the `age` property is declared as mutable using `var`, meaning that
the value of `age` can change later on.
Both properties are initialized directly when an object of `Person` [is created](#creating-objects-of-classes).

> By declaring class properties in the primary constructor using `val` or `var`, you make them accessible to 

Reviewer: Also, after taking a look at the feedback, if we still want to make this page more introductory, I think we should add an explicit example demonstrating the difference between

```
class Person(name: String)
// vs
class Person(val name: String)
```

aka the difference between primary constructor parameters and primary constructor properties.

> a parameterless constructor that uses those default values. 
> 
> This makes it easier to use Kotlin with libraries such as [Jackson](https://github.com/FasterXML/jackson) 
> or [Spring Data JPA](https://spring.io/projects/spring-data-jpa), which create class objects through 
> parameterless constructors.
>
> In the following example, Kotlin implicitly provides a parameterless constructor `Person()` that uses the default value `""`:
> 
> ```kotlin
> class Person(val personName: String = "")
> ```
>
{style="note"}

## Creating objects of classes

Like in other object-oriented programming languages, a class in Kotlin serves as a blueprint or template 
that defines the properties (attributes) and behaviors (functions) of objects. When you create an instance 
of a class, you are creating a concrete object based on that blueprint.

To create an object of a class, use the class name followed by parentheses (`()`), similar to calling a function:

```kotlin
// Creates an object of the 'Person' class
val person1 = Person()
```

In Kotlin, you can create objects:

* **Without arguments** (`()`): creates an object using a constructor with default values.

> * **With arguments** (`("value")`): creates an object by passing specific values to the constructor.

Reviewer: ```suggestion
* **With arguments** (`(value)`): creates an object by passing specific values to the constructor.
```

> As you can see, you can assign the created object to a mutable (`var`) or immutable (`val`) [variable](basic-syntax.md#variables):

Reviewer: ```suggestion
As you can see, you can assign the created object to a mutable (`var`) or read-only (`val`) [variable](basic-syntax.md#variables):
```

```kotlin
// Creates an object by passing a specific value and assigns it to a mutable variable
var person2 = Person("Joe")

> // Creates an object using the default constructor's value and assigns it to an immutable variable

Reviewer: ```suggestion
// Creates an object using the default constructor's value and assigns it to a read-only variable
```

val person1 = Person()
```

> You can create objects wherever you need them, inside the `main()` function, within other functions, or as part of a class. 

Reviewer: What do you mean when you say `as part of a class`?

When creating objects inside the `main()` function, when the program runs, execution starts from `main()`. 
Here is where you instantiate objects to interact with them. When creating objects of a class, you're invoking constructors 

> (for example, `Person()` or `Person("Joe")`) directly inside the `main()` function.

Reviewer: I feel that, for some reason, here we begin talking about smth unrelated to classes, i.e., about the purposes of the `main` function, when it is invoked, etc. Maybe we should create a separate page about it then, and simply reference it here?

Let's look at an example of how object creation works.

The following code defines a `Person` class with a property for storing a name. 
When an object of the class is created, an initialization block runs to print a message. The example demonstrates 
creating an object with both the default constructor's value and a specific value:

```kotlin
// Class header with primary constructor that initializes 'name' with a default value
class Person(val name: String = "Sebastian")

fun main() {
    // Creates an object using the default constructor's value
    val person1 = Person()
    println("Person object created: ${person1.name}")
    // Person object created: Sebastian

    // Creates an object by passing a specific value
    val person2 = Person("Joe")
    println("Person object created: ${person2.name}")
    // Person object created: Joe
}
```
{kotlin-runnable="true"}

> > In Kotlin, unlike other object-oriented programming languages like Java, 

Reviewer: Nit: "unlike ... like" reads weird

> there is no need for the `new` keyword when creating objects of a class.
>
{style="note"}

> Additionally, you can create objects inside another function and call that function from `main()`.

Reviewer: I think we've already said that before, and repeating the same fact again feels weird.

see the [Nested classes](nested-classes.md) section.

## Inheritance

Class inheritance in Kotlin allows you to create a new class (derived class) from an existing class (base class), 
inheriting its properties and functions while adding or modifying behavior.

> For detailed information about inheritance hierarchies and the use of the `open` keyword, see the [Inheritance](inheritance.md) section.
> 
{style="note"}

## Abstract classes

An abstract class in Kotlin is a class that cannot be instantiated directly and is designed to be inherited 
by other classes. 

An abstract class can define abstract properties and functions, which must be implemented 
by subclasses.

You can declare an abstract class using the `abstract` modifier:

```kotlin
abstract class Person
```

Like any other class, abstract classes can also have constructors. 
These constructors initialize class properties and enforce required parameters for subclasses:

```kotlin
abstract class Person(val name: String, val age: Int)
```

An abstract class can have both abstract and non-abstract members (properties and functions). 
To declare a member as abstract, you must use the `abstract` keyword explicitly.

> Abstract members do not have an implementation in the abstract class. The `override`

Reviewer: I think we should add a sentence about how one provides an implementation by writing an override in an inheritor. That will also bridge us to the `override` keyword which is used for overrides.

keyword ensures that a subclass provides an actual implementation for an abstract function:

```kotlin
// Abstract class with a primary constructor that defines 'name' and 'age'
abstract class Person(
    val name: String,
    val age: Int
) {

>     // Abstract member (doesn't provide default implementation, must be implemented by subclasses)

Reviewer: ```suggestion
    // Abstract member (doesn't provide implementation, must be implemented by subclasses)
```

Reviewer: It's a bit more precise to talk about base (not default) implementations, as "default" term is already used in Kotlin for other things.

    abstract fun introduce()

>     // Non-abstract member (has a default implementation)

Reviewer: ```suggestion
    // Non-abstract member (has an implementation)
```

        println("Hello, my name is $name.")
        // Hello, my name is Alice.
    }
}

// Subclass that provides an implementation for the abstract member
class Student(
    name: String,
    age: Int,
    val school: String
) : Person(name, age) {
    override fun introduce() {
        println("I am $name, $age years old, and I study at $school.")
        // I am Alice, 20 years old, and I study at Engineering University.
    }
}

fun main() {
    // Creates an object of the Student class
    val student = Student("Alice", 20, "Engineering University")
    // Calls the non-abstract member
    student.greet()
    // Calls the overridden abstract member
    student.introduce()
}
```
{kotlin-runnable="true"}

You don't need to annotate abstract classes or functions with the `open` keyword because they are implicitly 
open and designed to be inherited. 

[For more details about the `open` keyword, see Inheritance](inheritance.md#open-keyword).

## Companion objects

In Kotlin, a [companion object](object-declarations.md#companion-objects) is a type of object declaration 
that allows you to access its members using the class name without needing an object from the class.

> If you need to write a function that can be called without creating an object of a class but still needs access to the 

Reviewer: I would rephrase this to smth like "If you need to write a function that can be called without creating an object of a class but is still logically connected to the class (such as a factory function), you can define ..."

> class's companion object members, you can define it inside an [object declaration](object-declarations.md) within the class:

Reviewer: ```suggestion
class's companion object members, you can define it inside an companion [object declaration](object-declarations.md) within the class:
```

[//]: # (title: Classes)

Like other object-oriented languages, Kotlin uses _classes_ to encapsulate data and behavior 

> for reusable, structured code.

Reviewer: I would also prefer to see a very simple example in the beginning of "class -> ctor -> created object instance" which we then iteratively expand on, by adding more properties, functions, ctors, init blocks or other things.

```kotlin
// Class with a primary constructor that initializes 'name' and 'age'
class Person(val name: String, var age: Int) {
    init {
        // Initializer block executed when an object is created
        println("Person created: $name, age $age")
    }
}
```

In the example above, after the class declaration, there's an initializer block (`init{}`) containing runnable code 
to print a message that uses the class properties.

Considering the same code example, if you [create an object](#creating-objects-of-classes) of the `Person` class
by passing values for `name` and `age`, the output is:

```kotlin
//sampleStart
// Class with a primary constructor that initializes 'name' and 'age'
class Person(val name: String, var age: Int) {
    init {
        // Initializer block executed when an object is created
        println("Person created: $name, age $age.")
        // Person created: John, age 30.
    }
}

fun main() {
    // Creates an object of the 'Person' class
    Person("John", 30)
}
//sampleEnd
```
{kotlin-runnable="true"}

You can add as many initializer blocks as you need, and they will be executed in the order in which they appear in the class body, 

> interleaved with property initializers (`init{}`).

Author: Initializer block, I placed it where it corresponds.

// Class with a primary constructor that initializes 'name' and 'age'
class Person(val name: String, var age: Int) {
    // First initializer block
    init {
        // Runs first when an object is created
        println("Person created: $name, age $age.")
        // Person created: John, age 30.
    }

    // Second initializer block
    init {
        // Runs after the first initializer block
        if (age < 18) {
            println("$name is a minor.")
        } else {
            println("$name is an adult.")
            // John is an adult.
        }
    }
}

fun main() {
    // Creates an object of the 'Person' class
    Person("John", 30)
}
//sampleEnd
```
{kotlin-runnable="true"}

You can use the primary constructor parameters in the initializer blocks. For example, in the code above, the first and second initializers use
the `name` and `age` parameters from the primary constructor. 

### Secondary constructors

In Kotlin, a secondary constructor is an additional constructor that a class can have beyond its primary constructor. 
Secondary constructors allow you to provide different ways to initialize an object when the primary constructor alone 
isn't enough.

> Secondary constructors are useful when interacting with Java code that requires traditional constructors or if

Author: Is it better?

    val name: String,
    var age: Int
) {
    // Secondary constructor with direct delegation to the primary constructor
    constructor(name: String) : this(name, 0) {
        println("Person created with default age: $age and name: $name.")
        // Person created with default age: 0 and name: Alice.
        // Person created with default age: 0 and name: Bob.
    }

    // Secondary constructor with indirect delegation: 'this("Bob")' -> 'constructor(name: String)' -> primary constructor
    constructor() : this("Bob") {
        println("New person created with default age: $age and name: $name.")
        // New person created with default age: 0 and name: Bob.
    }
}

fun main() {
    // Creates an object based on the direct delegation
    Person("Alice")

    // Creates an object based on the indirect delegation
    Person()
}
```
{kotlin-runnable="true"}

In classes with initializer blocks (`init {}`), the code within these blocks becomes part of the primary constructor.
When a secondary constructor is invoked, it delegates to the primary constructor first. As a result, all initializer blocks
and property initializers run before the body of the secondary constructor. 

> Whenever an [object of a class is created](#creating-objects-of-classes), all initializer blocks run before any secondary constructor logic.

Author: Do you mean:
First the full code, clean, without inline output comments.
Then separately show what the program prints when executed.

If you don't want your class to have a public constructor, declare an empty primary constructor with non-default visibility:

```kotlin
class Person private constructor() { /*...*/ }
```

> On the JVM, if all primary constructor parameters have default values, the compiler generates 
> an additional parameterless constructor that uses those default values. 
> 
> This makes it easier to use Kotlin with libraries such as [Jackson](https://github.com/FasterXML/jackson) 
> or [Spring Data JPA](https://spring.io/projects/spring-data-jpa), which create class objects through 
> parameterless constructors.
>
> In the following example, Kotlin automatically generates a parameterless constructor `Person()` that uses the default value `""`:
> 
> ```kotlin
> class Person(val personName: String = "")
> ```
>
{style="note"}

## Creating objects of classes

Like in other object-oriented programming languages, a class in Kotlin serves as a blueprint or template 
that defines the properties (attributes) and behaviors (functions) of objects. When you create an instance 
of a class, you are creating a concrete object based on that blueprint.

> To create an object of a class, use the class name followed by parentheses (`()`), similar to calling a function:

Reviewer: In the previous paragraphs we were already creating objects of classes using this syntax. Maybe we should briefly introduce how to call a constructor before, and here we are explaining that in more detail?

Author: Added it in primary constructor, right before the first runnable example of object creation.

```kotlin
// Creates an object of the 'Person' class
val person1 = Person()
```

In Kotlin, you can create objects:

* **Without arguments** (`()`): creates an object using a constructor with default values.
* **With arguments** (`("value")`): creates an object by passing specific values to the constructor.

As you can see, you can assign the created object to a mutable (`var`) or immutable (`val`) [variable](basic-syntax.md#variables):

```kotlin
// Creates an object by passing a specific value and assigns it to a mutable variable
var person2 = Person("Joe")

// Creates an object using the default constructor's value and assigns it to an immutable variable
val person1 = Person()
```

> Typically, you create objects within the `main()` function because it serves as the entry point of your Kotlin program. 

Reviewer: This statement is not true, you create objects whenever you need to do that, and not only within the `main` function

    val name: String,
    val age: Int
) {
    // Abstract member (doesn't provide default implementation, must be implemented by subclasses)
    abstract fun introduce()

    // Non-abstract member (has a default implementation)
    fun greet() {
        println("Hello, my name is $name.")
        // Hello, my name is Alice.
    }
}

// Subclass that provides an implementation for the abstract member
class Student(
    name: String,
    age: Int,
    val school: String
) : Person(name, age) {
    override fun introduce() {
        println("I am $name, $age years old, and I study at $school.")
        // I am Alice, 20 years old, and I study at Engineering University.
    }
}

fun main() {
    // Creates an object of the Student class
    val student = Student("Alice", 20, "Engineering University")
    // Calls the non-abstract member
    student.greet()
    // Calls the overridden abstract member
    student.introduce()
}
```
{kotlin-runnable="true"}

You don't need to annotate abstract classes or functions with the `open` keyword because they are implicitly 
open and designed to be inherited. 

Similarly, you can override a non-abstract `open` member with an abstract one:

> ## Open keyword

Reviewer: I would say this section makes more sense in the `Inheritance` part of the documentation

) : Person(name) {
    override fun introduce() {
        println("Hi, I'm $name, and I study at $school.")
    }
}
```

In addition, you need to add the `open` keyword to every property
or function that can be overridden.

If you override a member of a base class, the overriding member
will also be open by default. If you want to change this and forbid the subclasses of your
class from overriding your implementation, you can explicitly mark the overriding
member as `final`:

```kotlin
// Base class with the 'open' keyword to allow inheritance
open class Person(val name: String) {
    // Open function that can be overridden in a subclass
    open fun introduce() {
        println("Hello, my name is $name.")
    }
}

// Subclass inheriting from 'Person' and overriding the 'introduce()' function
class Student(name: String, val school: String) : Person(name) {
    // The 'final' keyword prevents further overriding in subclasses
    final override fun introduce() {
        println("Hi, I'm $name, and I study at $school.")
    }
}
```

You don't need to annotate abstract classes or functions with the `open` keyword because they are implicitly 
open and designed to be inherited. 

> Similarly, you can override a non-abstract `open` member with an abstract one:

Reviewer: Tbh, I do not think this example adds anything useful for one reading the documentation. Such pattern of overriding `open` with `abstract` is very strange and should not be used unless there are very specific circumstances.

```kotlin
// Abstract class with a primary constructor that defines 'name'
abstract class Person(val name: String) {
  // Non-abstract open member that allows overriding
  open fun greet() {
    println("Hello, my name is $name.")
  }

  // Abstract member
  abstract fun introduce()
}

// Subclass that overrides the 'greet()' function with an abstract one
abstract class Student(name: String, val school: String) : Person(name) {
  // Overrides the non-abstract 'greet()' function with an abstract one
  abstract override fun greet()

  // Overrides the abstract 'introduce()' function with an abstract one
  override fun introduce() {
    println("I am $name, and I study at $school.")
  }
}

```

## Companion objects

In Kotlin, a [companion object](object-declarations.md#companion-objects) is a type of object declaration 
that allows you to access its members using the class name without needing an object from the class.

If you need to write a function that can be called without having a class object but still needs access 

> to the internals of the class, you can define it inside `companion object` within that class:

Reviewer: How do you define "internals of the class" here? A companion object can access its own internals, but class itself does not have any internals until you have created an instance of it.

Reviewer: Also, here the terminology begins being a bit confusing, as "class object" and "companion object" look very similar, but mean significantly different things

  * [Secondary constructors](#secondary-constructors)
  * [Initializer blocks](#initializer-blocks)
  * [Functions](functions.md)
  * [Properties](properties.md)
  * [Nested and inner classes](nested-classes.md)
  * [Object declarations](object-declarations.md)

Both the class header and class body can be minimal. 
If the class does not have a body, you can omit the curly braces `{}`:

```kotlin
// Class with name and age properties declared in the primary constructor, but without a body 
class Person(val name: String, var age: Int)
```

Here's an example of a class with a header, body, and an [instance created](#creating-instances-of-classes) from it:

```kotlin
// Person class with a primary constructor that initializes the name property
class Person(val name: String) {
    // Class body with age property
    var age: Int = 0
}

fun main() {
    // Creates an instance of the Person class by calling the constructor
    val person = Person("Alice")

    // Accesses the instance's properties
    println(person.name)
    // Alice
    println(person.age)
    // 0
}
```
{kotlin-runnable="true"}

> ## Creating instances of classes

Reviewer: Maybe introduce constructors first and then talk about calling them?

Author: This is the issue why the previous version seemed too packed with information and code samples. In that version, all the runnable codes demonstrating instance creation required showed a concept that hadn’t been introduced yet, so it required explanation any time it appeared. In the current structure, creating an instance is explained before all the code samples that demonstrate it.

Additionally, as you can see from the last paragraph, the content is arranged to avoid unnecessary breaks. As soon as constructors are mentioned, there’s a smooth transition to the corresponding explanation. I tried different structures, and this one provides the most natural flow while avoiding repetition.

```kotlin
class Person(val name: String = "John", var age: Int = 30) { /*...*/ }
```

If no value is passed during [instance creation](#creating-instances-of-classes), the property uses the default value:

```kotlin
// Class with a primary constructor including default values for name and age
class Person(val name: String = "John", var age: Int = 30)

fun main() {
    // Creates an instance using default values
    val person = Person()
    println("Name: ${person.name}, Age: ${person.age}")
    // Name: John, Age: 30
}
```
{kotlin-runnable="true"}

Use the primary constructor parameters to initialize additional class properties directly in the class body:

```kotlin
// Class with a primary constructor including default values for name and age
class Person(
    val name: String = "John",
    var age: Int = 30
) {
    // description property initialized in the class body using the primary constructor parameters
    val description: String = "Name: $name, Age: $age"
}

fun main() {
    // Creates an instance of the Person class
    val person = Person()
    // Accesses the description property
    println(person.description)
    // Name: John, Age: 30
}
```
{kotlin-runnable="true"}

> Additionally, you can use a [trailing comma](coding-conventions.md#trailing-commas) in constructor declarations:

Reviewer: Maybe "As with functions, you can ..."?

```kotlin
class Person(
    val name: String,
    val lastName: String,
    var age: Int,
) { /*...*/ }
```

### Initializer blocks

> The primary constructor can't contain runnable code. Its purpose is to initialize the class and set its
> properties. If you want to run some code during [instance creation](#creating-instances-of-classes), use _initializer blocks_ inside the class body.

Reviewer: ```suggestion
The purpose of the primary constructor is to initialize the class and set its
properties, and usually you can achieve this without needing to run complex code blocks.
If you do need to do this in a primary constructor during [instance creation](#creating-instances-of-classes), use _initializer blocks_ inside the class body.
```

Reviewer: Also, maybe add that quite often `init` blocks are used for data validation, e.g., by calling `require` function?

```
class Person(val age: Int) {
    init { require(age > 0, "age must be positive") }
}
```

isn't enough.

Secondary constructors are useful for [Java interoperability](java-to-kotlin-interop.md) and when
you need multiple ways to initialize a class.

To declare a secondary constructor, use the `constructor`
keyword inside the class body with the constructor parameters within parentheses. Use curly braces with the constructor logic inside:

```kotlin
// Class header with a primary constructor that initializes name and age
class Person(val name: String, var age: Int) {
    // Secondary constructor that takes a String age and converts it to an integer
    constructor(name: String, age: String) : this(name, age.toIntOrNull() ?: 0) {
        println("$name created with converted age: $age")
        // Bob created with converted age: 8
    }
}

fun main() {
    // Uses the secondary constructor with a String age
    Person("Bob", "8")
}
```
{kotlin-runnable="true"}

> The expression `age.toIntOrNull() ?: 0` converts the `age` string to an integer. For more information about this expression, see [Null safety](null-safety.md).
>
{style="tip"}

In the code above, the secondary constructor delegates to the primary constructor via the `this` keyword, 

> passing `name` and the default `age` value.

Reviewer: "...and the converted to integer `age` value"

[//]: # (title: Classes)

Like other object-oriented languages, Kotlin uses _classes_ to encapsulate data (properties) and behavior (functions)
for reusable, structured code.

Classes are blueprints or templates for objects, which you 
create via [constructors](#constructors-and-initializer-blocks). 
When you [create an instance of a class](#creating-instances-of-classes), you are creating 
a concrete object based on that blueprint.

Kotlin offers a concise syntax for declaring classes. To declare a class, use the `class` keyword 
followed by the class name:

```kotlin
class Person { /*...*/ }
```

The class declaration consists of: 
* **Class header**, including but not limited to:
  * `class` keyword
  * Class name
  * Type parameters (if any)
  * [Primary constructor](#primary-constructor) (optional)
* **Class body** (optional), surrounded by curly braces `{}`, and including **class members** such as:
  * [Secondary constructors](#secondary-constructors)
  * [Initializer blocks](#initializer-blocks)
  * [Functions](functions.md)
  * [Properties](properties.md)
  * [Nested and inner classes](nested-classes.md)
  * [Object declarations](object-declarations.md)

> Both the class header and class body can be minimal. 

Author: It means that the class header and the class body can be reduced to their simplest possible form, containing only what is strictly necessary. Class header: just the class name. Class body: just a few required members, or even completely absent. 

Do you think it's clearer if we explain it a bit more?

// and assigns it to a mutable variable
var person1 = Person()

// Creates an instance by passing a specific value 
// and assigns it to a read-only variable
val person2 = Person("Joe")
```

It's possible to create instances wherever you need them, inside the [`main()` function](basic-syntax.md#program-entry-point), within other functions, or inside another class.
Additionally, you can create instances inside another function and call that function from `main()`.

The following code declares a `Person` class with a property for storing a name. 
It also demonstrates
how to create an instance with both the default constructor's value and a specific value:

```kotlin
// Class header with a primary constructor 
// that initializes name with a default value
class Person(val name: String = "Sebastian")

fun main() {
    // Creates an instance using the default constructor's value
    val person1 = Person()

    // Creates an instance by passing a specific value
    val person2 = Person("Joe")

    // Accesses the instances' name property
    println(person1.name)
    // Sebastian
    println(person2.name)
    // Joe
}
```
{kotlin-runnable="true"}

> > In Kotlin, unlike other object-oriented programming languages such as Java,

Reviewer 2: ```suggestion
> In Kotlin, unlike other object-oriented programming languages,
```
To make it more language-agnostic :)

    // Name: John, Age: 30
}
```
{kotlin-runnable="true"}

As with functions, you can use [trailing comma](coding-conventions.md#trailing-commas) in constructor declarations:

```kotlin
class Person(
    val name: String,
    val lastName: String,
    var age: Int,
) { /*...*/ }
```

### Initializer blocks

The purpose of the primary constructor is to initialize the class and set its
properties. In most cases, this initialization doesn’t require complex code.

If you need to run complex code blocks when the primary constructor is executed during [instance creation](#creating-instances-of-classes), 
use _initializer blocks_ inside the class body.

Declare initializer blocks with the `init` keyword followed by curly braces `{}`. 
Write within the curly braces any code that you want to run during initialization:

```kotlin
// Class with a primary constructor that initializes name and age
class Person(val name: String, var age: Int) {
    init {
        // Initializer block executed when an instance is created
        println("Person created: $name, age $age.")
        // Person created: John, age 30.
    }
}

fun main() {
    // Creates an instance of the Person class

>     Person("John", 30)

Reviewer 2: ```suggestion
    Person("John", 30)
    // Person created: John, age 30.
```

```kotlin
class Person(val age: Int) {
    init {
        require(age > 0, "age must be positive")
    }
}
```

### Secondary constructors

In Kotlin, a secondary constructor is an additional constructor that a class can have beyond its primary constructor. 
Secondary constructors allow you to provide different ways to initialize an instance when the primary constructor alone 
isn't enough.

Secondary constructors are useful for [Java interoperability](java-to-kotlin-interop.md) and when
you need multiple ways to initialize a class.

To declare a secondary constructor, use the `constructor`
keyword inside the class body with the constructor parameters within parentheses. Use curly braces with the constructor logic inside:

```kotlin
// Class header with a primary constructor that initializes name and age
class Person(val name: String, var age: Int) {

    // Secondary constructor that takes a String age
    // and converts it to an integer
    constructor(name: String, age: String) : this(name, age.toIntOrNull() ?: 0) {
        println("$name created with converted age: $age")

>         // Bob created with converted age: 8

Reviewer 2: ```suggestion
```

    }
}

fun main() {
    // Uses the secondary constructor with a String age

>     Person("Bob", "8")

Reviewer 2: ```suggestion
    Person("Bob", "8")
    // Bob created with converted age: 8
```
I think it makes more sense to add the printed statement here for the constructor.

>
{style="tip"}

In the code above, the secondary constructor delegates to the primary constructor via the `this` keyword, 
passing `name` and the `age` value converted to an integer.

In Kotlin, secondary constructors must delegate to the primary constructor. This delegation ensures that all primary 
constructor initialization logic is executed before any secondary constructor logic runs. 

Constructor delegation can be: 
* **Direct**, where the secondary constructor calls the primary constructor immediately.
* **Indirect**, where one secondary constructor calls another, which in turn delegates to the primary constructor.

Here's an example demonstrating how direct and indirect delegation works:

```kotlin
// Class header with a primary constructor that initializes name and age
class Person(
    val name: String,
    var age: Int
) {
    // Secondary constructor with direct delegation 
    // to the primary constructor
    constructor(name: String) : this(name, 0) {
        println("Person created with default age: $age and name: $name.")

>         // Person created with default age: 0 and name: Alice.

Reviewer 2: Can you move the code comments to where they will appear when triggered by the main function?

open and designed to be inherited. 

[For more details about the `open` keyword, see Inheritance](inheritance.md#open-keyword).

## Companion objects

In Kotlin, a [companion object](object-declarations.md#companion-objects) is a type of object declaration 
that allows you to access its members using the class name without needing an object from the class.

Suppose you need to write a function that can be called without creating an instance of a class, but it is still logically 
connected to the class (such as a factory function). In that case, you can declare it inside a companion [object declaration](object-declarations.md) within the class:

```kotlin
// Class with a primary constructor that declares the name property
class Person(
    val name: String
) {
    // Class body with a companion object
    companion object {
        fun createAnonymous() = Person("Anonymous")
    }
}

fun main() {
    // Calls the function without creating an instance of the class
    val anonymous = Person.createAnonymous()
    println(anonymous.name)
    // Anonymous
}
```
{kotlin-runnable="true"}

> If you declare `companion object` inside your class,

Reviewer 2: ```suggestion
If you declare a `companion object` inside your class,
```
Does "companion object" need to be in inlined code format here?
